src/server/frontend_wayland/ext_image_capture_v1.cpp

```diff
@@ -0,0 +1,607 @@
+            output_space_area,
+            transform,
+            overlay_cursor,
+            buffer_size,
+            {{}, buffer_size});
+        break;
+    }
+
+    damage_amount = DamageAmount::none;
+}
+
+void mf::ExtImageCopyCaptureSessionV1::create_frame(
+    wl_resource* new_resource)
+{
+    if (current_frame)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::duplicate_frame,
+            "A frame already exists for this session"));
+    }
+    current_frame = wayland::make_weak(
+        new ExtImageCopyCaptureFrameV1{new_resource, this, ctx});
+}
+
+mf::ExtImageCopyCaptureFrameV1::ExtImageCopyCaptureFrameV1(
+    wl_resource *resource,
+    ExtImageCopyCaptureSessionV1* session,
+    std::shared_ptr<ExtImageCaptureV1Ctx> const& ctx)
+    : wayland::ImageCopyCaptureFrameV1{resource, Version<1>()},
+      session{session},
+      ctx{ctx}
+{
+}
+
+void mf::ExtImageCopyCaptureFrameV1::attach_buffer(
+    [[maybe_unused]] wl_resource* buffer)
+{
+    if (capture_has_been_called)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::already_captured,
+            "Cannot attach buffer after capture"));
+    }
+
+    auto buf = wayland::Buffer::from(buffer);
+    if (!buf)
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error(
+            "wl_buffer@" +
+            std::to_string(wl_resource_get_id(buffer)) +
+            " is not a wayland::Buffer"));
+    }
+    target = wayland::make_weak(buf);
+}
+
+void mf::ExtImageCopyCaptureFrameV1::damage_buffer(
+    int32_t x,
+    int32_t y,
+    int32_t width,
+    int32_t height)
+{
+    if (!session)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::invalid_buffer_damage,
+            "session is gone"));
+    }
+    // TODO: this needs to be converted to output space coordinates
+    session.value().apply_damage(geom::Rectangle{{x, y}, {width, height}});
+}
+
+void mf::ExtImageCopyCaptureFrameV1::capture()
+{
+    if (capture_has_been_called)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::already_captured,
+            "capture already called"));
+    }
+    if (!target)
+    {
+        BOOST_THROW_EXCEPTION(wayland::ProtocolError(
+            resource, Error::no_buffer,
+            "no buffer attached"));
+    }
+    capture_has_been_called = true;
+
+    if (!session)
+    {
+        send_failed_event(FailureReason::stopped);
+        return;
+    }
+    session.value().maybe_capture_frame();
+}
+
+bool mf::ExtImageCopyCaptureFrameV1::is_ready() const
+{
+    return capture_has_been_called;
+}
+
+void mf::ExtImageCopyCaptureFrameV1::capture_frame(
+    geom::Rectangle output_space_area,
+    glm::mat2 transform,
+    bool overlay_cursor,
+    geom::Size buffer_size,
+    geom::Rectangle buffer_space_damage)
+{
+    if (!session)
+    {
+        send_failed_event(FailureReason::stopped);
```

Comment:
We *might* need to revisit this later; I don't think the protocol actually prohibits destroying the session before the frame has been captured (although that would be odd).